File: sdk/typescript/src/zip.ts
import { unzipSync } from "fflate";
import type { ExtractResult, Image } from "./models.js";
import { createEmptyResult } from "./models.js";

const IMAGE_EXTENSIONS = new Set([
  ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".svg", ".webp",
]);

function extname(filename: string): string {
  const dot = filename.lastIndexOf(".");
  return dot === -1 ? "" : filename.slice(dot).toLowerCase();
}

function basename(filepath: string): string {
  const parts = filepath.replace(/\\/g, "/").split("/");
  return parts[parts.length - 1] ?? "";
}

export function parseZip(
  zipBytes: Uint8Array,
  taskId: string,
  filename: string | null = null,
): ExtractResult {
  const result = createEmptyResult(taskId, "done");
  result.filename = filename;
  result._zipBytes = zipBytes;

  const entries = unzipSync(zipBytes);
  const images: Image[] = [];
  let contentList: Record<string, unknown>[] | null = null;

  for (const [relPath, data] of Object.entries(entries)) {
    if (relPath.endsWith("/")) continue;

    const name = basename(relPath);
    const ext = extname(name);
    const text = () => new TextDecoder().decode(data);

    if (ext === ".md") {
      result.markdown = text();
    } else if (
      name.endsWith("_content_list.json") ||
      name === "content_list.json"
    ) {
      contentList = JSON.parse(text()) as Record<string, unknown>[];
    } else if (ext === ".json" && contentList == null) {
      try {
        const parsed: unknown = JSON.parse(text());
        if (Array.isArray(parsed)) {
          contentList = parsed as Record<string, unknown>[];
        }
      } catch {
        // not a valid JSON array — skip
      }
    } else if (IMAGE_EXTENSIONS.has(ext)) {
      images.push({ name, data: new Uint8Array(data), path: relPath });
    } else if (ext === ".docx") {
      result.docx = new Uint8Array(data);
    } else if (ext === ".html" || ext === ".htm") {
      result.html = text();
    } else if (ext === ".tex") {
      result.latex = text();
    }
  }

  result.contentList = contentList;
  result.images = images;
  return result;
}

File: sdk/go/zip.go
package mineru

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"io"
	"path"
	"strings"
)

// parseZip extracts content from a MinerU result zip into an ExtractResult.
func parseZip(zipBytes []byte, taskID, filename string) (*ExtractResult, error) {
	zr, err := zip.NewReader(bytes.NewReader(zipBytes), int64(len(zipBytes)))
	if err != nil {
		return nil, err
	}

	r := &ExtractResult{
		TaskID:   taskID,
		State:    "done",
		Filename: filename,
		zipBytes: zipBytes,
	}

	for _, f := range zr.File {
		if f.FileInfo().IsDir() {
			continue
		}
		name := path.Base(f.Name)
		ext := strings.ToLower(path.Ext(name))

		data, err := readZipFile(f)
		if err != nil {
			return nil, err
		}

		switch {
		case ext == ".md":
			r.Markdown = string(data)

		case strings.HasSuffix(name, "_content_list.json") || name == "content_list.json":
			var cl []map[string]any
			if json.Unmarshal(data, &cl) == nil {
				r.ContentList = cl
			}

		case ext == ".json" && r.ContentList == nil:
			var cl []map[string]any
			if json.Unmarshal(data, &cl) == nil {
				r.ContentList = cl
			}

		case isImageExt(ext):
			r.Images = append(r.Images, Image{Name: name, Data: data, Path: f.Name})

		case ext == ".docx":
			r.Docx = data

		case ext == ".html" || ext == ".htm":
			r.HTML = string(data)

		case ext == ".tex":
			r.LaTeX = string(data)
		}
	}
	return r, nil
}

func readZipFile(f *zip.File) ([]byte, error) {
	rc, err := f.Open()
	if err != nil {
		return nil, err
	}
	defer rc.Close()
	return io.ReadAll(rc)
}

func isImageExt(ext string) bool {
	switch ext {
	case ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".svg", ".webp":
		return true
	}
	return false
}
